tests/test2.nim

```diff
@@ -0,0 +1,226 @@
+      succeed
+    else:
+      discard
+
+     
+    # block match0:
+    #   var error0: MatchingError
+    #   block branch0:
+    #     if not (container.kind == HContainer):
+    #       error0 = ..
+    #       break branch0
+    #     if not (container.c.kind == CSeq):
+    #       error0 = ..
+    #       break branch0
+    #     if not (container.c.elements.len == 0): # elements[0] is true
+    #       error0 = ..
+    #       break branch0
+    #     let value = container.c.elements[0]
+    #     echo value
+    #     check value.kind == HInt and value.i == 0
+    #     succeed
+    #     break match0
+    #   discard
+
+  matchtest "match variant field":
+    var a = A(kind: HInt, i: 0)
+
+    a.match:
+    of HInt(i: `i`):
+      check i == 0
+      succeed
+    else:
+      discard
+
+
+  matchtest "match variant literal":
+    A.matchLit int, HInt
+    A.matchLit string, HString
+
+    var a = A(kind: HInt, i: 0)
+    var b = A(kind: HString, s: "e")
+
+    a.match:
+    of 0:
+      succeed
+    else:
+      discard
+
+    # block match0:
+    #   var error0: MatchingError
+    #   block branch0:
+    #     if not (a.kind == HInt and a.i == 0):
+    #       error0 = ..
+    #       break branch0
+    #     succeed
+    #     break match0
+    #   discard
+
+    b.match:
+    of "f":
+      discard
+    of "e":
+      succeed
+    else:
+      discard
+
+    # block match0:
+    #   var error0: MatchingError
+    #   block branch0:
+    #     if not (a.kind == HString and a.s == "f"):
+    #       error0 = ..
+    #       break branch0
+    #     break match0
+    #   var error1: MatchingError
+    #   block branch1:
+    #     if not (a.kind == HString and a.s == "e"):
+    #       error1 = ..
+    #       break branch1
+    #     succeed
+    #     break match0
+    #   discard
+
+  matchtest "match variant args with auto":
+    Container.matchArgs elements
+
+    var a = Container(kind: CSeq, elements: @[A(kind: HInt, i: 2), A(kind: HInt, i: 4)])
+    a.match:
+    of CSeq(HInt(2), HInt(4)):
+      succeed
+    else:
+      discard
+
+  matchtest "match normal objects":
+    type
+      Obj = object
+        a*: int
+        b*: int
+
+    var a = Obj(a: 0, b: 0)
+    a.match:
+    of Obj(a: `a`, b: 0):
+      check a == 0
+      succeed
+    else:
+      discard
+
+  matchtest "match seq":
+    var a = @[0, 1, 2]
+    a.match:
+    of @[0]:
+      discard
+    of @[0, 1, `last`]:
+      check last == 2
+      succeed
+    else:
+      discard
+
+    # block match0:
+    #   var error0: ..
+    #   block branch0:
+    #     if not (a.len == 1 and a[0] == 0):
+    #       error0 = ..
+    #       break branch0
+    #   var error1: ..
+    #   block branch1:
+    #     if not (a.len == 3 and a[0] == 0 and a[1] == 1):
+    #       error1 = ..
+    #       break branch1
+    #     let last = a[2]
+    #     check last == 2
+    #     succeed
+
+    a.match:
+    of @[`first`]:
+      discard
+    of @[0, *`last`]:
```

Comment:
We discussed and you agreed that you would drop this. The problem I have with this here is that this cannot be implemented like you suggested without creating a copy of the seq. I do not want that in the pattern matching expression any allocation on the heap is happening. Also mutating the copy would not properly propaget the changes to the original seq. But if you are able to force `last` to be an `openArray` that actually references the original seqence, I would accept it. But I am not sure if the nim compiler allows you to do that.

Reply:
Well, how would you match 
```
type 
  B = object
    a: seq[int]

of B(a = `list`)
```
you'll need to assign a sequence to list, which will create a copy. 
I think the "no allocation" wish is very nice, but very limiting: if demands one to always work with compile time known sizes. And I already hit the limits of that very quickly (I tagged you in #nim )
I can't match `nnkCall` with unknown number of args, I can't express that I want all the children of a node to share the same shape, I can't do anything with variadic number of elements, and a lot of the ast nodes are such variadic collections

Reply:
so I didn't use the ast match macro, because most of my logic wasn't expressable in it: very often I need to collect sequences of subparts of nodes, and without way to express that, I still have to write a lot of boilerplate

Reply:
"Also, 'would not propagate the changes', how does that work now?"
It doesn't. The ast pattern matching is at it's current state a minimal viable product. I just don't want to add any featurs that make general references into the original tree impossible.

Reply:
and for you use cases with `*`. Show them to me and I show you how you can match them without `*`.

Reply:
```nim
macro f(node: untyped): untyped =
  # f(typename, arg1, arg2..)
  node.matchAst:
  of nnkCall(`typename` @ nnkIdent, `args` @ *nnkIntLit): # I want to somehow reference part of the children and confirm they all share the same shape
  ..
```

```nim
node.matchAst:
of nnkObjConstr(`componentName`, nnkExprColonExpr(`names[]` @ nnkIdent, `values[]`)):
```
I want to check the internal shapes of the args and extract directly stuff from them without going through several loops (the last breaks the "reference the original tree", but 90% of the time when I do this, I don't care about that: I just want to extract something from the input with powerful syntax and create new NimNode output)

Reply:
I am sure I can think of many if I look at my macros: basically for each NimNode that has variable number of children, I probably need it

Reply:
The problem is, I can't look in your head. I need examples.

Reply:
I literally gave you two examples ^

Reply:
the first one is explained, the second one I want to match f A(a: v, b: w) and turn it into

```nim
block:
  let c = stuff with A
  c.a = ..
  c.b = 
  c
```

Reply:
sorry for the lack of context

Reply:
Sorry, for some reason github thought it was a good idea to fold the conversation and your examples were just invisible. 

The first example is a valid example, I agree on that. The second one though is a bit weird to me. I would rather use a pattern like this:

```nim
of nnkObjConstr(`componentName`, `stuff` @ *nnkExprColonExpr(`name` @ nnkIdent, `value`)):
    stuff.len
    stuff[0].name
    stuff[0].value
    stuff[1].name
    ...
```

Reply:
yeah I agree, the second one with `stuff` is a pretty good alternative.
I agree that it might be not easy, but I have to think a bit more about it

Reply:
well it is kind of hard, because ``stuff`` has to be a collection of ``NimNode``. Otherwise you couldn't just use ``stuff`` anymore.

Reply:
that's not hard, you can have it and a locally defined template `name` : `node[0]`